### ISSUE TITLE
Allow UVA sampling where train_nid/train_eid is provided on the CPU

### ISSUE THREAD

**Issue by nv-dlasalle** [dmlc/dgl#3855]
## 🚀 Feature
For many sampling techniques, the number of input nodes is much larger than the number of output nodes. As a result the cost of copying the output_nodes in a mini-batch to the GPU is relatively negligible (especially since intermediate layer seed nodes will be produced on the GPU in UVA sampling).

If `use_uva=True` in the dataloader, it would be great if it would then copy the output_nodes for the  mini-batch to the GPU before calling sampling. This would simplify the use_ddp workflow, as users could just store their train_nid in shared memory, and not duplicate them across all GPUs.

**Comment by BarclayII** [comment]
I don't quite get it :(  Could you give an example showing the problem?

**Comment by nv-dlasalle** [comment]
Right now, to use UVA sampling, a user must copy train_id to the GPU:
```
train_id = train_id.to(dev_id)

dataloader.DataLoader(g, train_id, ...)
```

For multi-GPU runs, this is even trickier, as the user must manually partition the ids before transferring them to the GPU, and do so in way that results in an even number of mini-batches:
```
block_size = (len(train_id)+world_size-1)//world_size
block_idx = torch.arange(block_size*rank, block_size*(rank+1)) % len(train_id)
train_id = train_id[block_idx]
train_id = train_id.to(dev_id)

dataloader.DataLoader(g, train_id, ..., use_uva=True, use_ddp=False)
```
This makes for much worse user experience when using UVA sampling. Their only alternative is to duplicate `train_id` on each GPU, which can use up a significant amount of memory, especially for unsupervised training. For all 1.6 billion edges in ogbn-papers100M, this would be 13GB per GPU if kept direct, and 26GB if made undirected.

Given that once we shuffle the ids for a given epoch, we only access relatively small contiguous blocks of the ids to build mini-batches from, I'm proposing we have the dataloader call `.to(device)` on the set ids used to create the mini-batch.
A rather naive implementation of this would be to simply change `__call__()` in the `CollateWrapper`:
```
diff --git a/python/dgl/dataloading/dataloader.py b/python/dgl/dataloading/dataloader.py
index 10852d9c..d2531129 100644
--- a/python/dgl/dataloading/dataloader.py
+++ b/python/dgl/dataloading/dataloader.py
@@ -523,11 +523,14 @@ class CollateWrapper(object):
     """Wraps a collate function with :func:`remove_parent_storage_columns` for serializing
     from PyTorch DataLoader workers.
     """
-    def __init__(self, sample_func, g):
+    def __init__(self, sample_func, g, item_device=None):
         self.sample_func = sample_func
         self.g = g
+        self.item_device = item_device

     def __call__(self, items):
+        if self.item_device:
+            items = F.copy_to(items, self.item_device) # no-op if already on self.item_device
         batch = self.sample_func(self.g, items)
         return recursive_apply(batch, remove_parent_storage_columns, self.g)

@@ -795,7 +798,8 @@ class DataLoader(torch.utils.data.DataLoader):

         super().__init__(
             self.dataset,
-            collate_fn=CollateWrapper(self.graph_sampler.sample, graph),
+            collate_fn=CollateWrapper(self.graph_sampler.sample, graph,
+                item_device=device if use_uva else None),
             batch_size=None,
             worker_init_fn=worker_init_fn,
             **kwargs)
```
Realistically there are other places in the code currently where we inspect the device of the train_ids that we would need to change, and there may be a better place to make this transfer.

**Comment by BarclayII** [comment]
Partitioning should be already handled in `DDPTensorizedDataset` class, so users don't need to partition it themselves - unless they are implementing their own dataset.

I'm OK with copying the ID to the GPUs in `CollateWrapper`.  The question is, should we make that the default behavior?  If under the case of full-GPU training I think that will cause a regression.

**Comment by nv-dlasalle** [comment]
The problem with `DDPTensorizedDataset` is that it requires duplicating the entire set of training ids on each GPU. So if we have 1 billion edge graph we're doing unsupervised training on, each GPU needs to store an 8GB tensor (1 billion 64-bit indices). This is why users can't use it to partition the dataset, when the training ids are on the GPU, as required by UVA sampling currently.

I'd suggest we only attempt the copy if `use_uva=True` to avoid introducing overheads (however I believe for most of the backends F.copy_to() is a no-op when the tensor is already on the specified device--so it shouldn't really matter).

**Comment by BarclayII** [comment]
> it requires duplicating the entire set of training ids on each GPU

Not necessarily for homogeneous graphs if we incorporate your changes to `CollateWrapper`.  The logic of `DDPTensorizedDataset` does not depend on having training IDs duplicated since what it does is only maintaining an index tensor for permutation (which is on shared CPU memory).

I saw the real problem is in heterogeneous graphs, where we are actually constructing a 2xN tensor with type ID and node ID stacked from a dictionary of node IDs.  We need a kernel that slices a dictionary out from a dictionary to avoid Python overhead if we don't want to construct such a tensor for every GPU.
